### DISCUSSION TITLE
Why parser generation degrades on optional recursive parts?

### DISCUSSION THREAD

**Discussion by susliko** [tree-sitter/tree-sitter#1948] [Q&A] [answered]
Consider an example: tree-sitter-scala grammar
Grammar from [the latest commit](https://github.com/tree-sitter/tree-sitter-scala/commit/140c96cf398693189d4e50f76d19ddfcd8a018f8) (on my machine) is processed in 9.23 seconds:
```bash
❯ time ts gen && du -h src/parser.c
tree-sitter gen  9,23s user 1,92s system 99% cpu 11,151 total
9,6M    src/parser.c
```
But if we remove an `optional` combinator around an [else block](https://github.com/tree-sitter/tree-sitter-scala/blob/140c96cf398693189d4e50f76d19ddfcd8a018f8/grammar.js#L575) of `if_expression`, we get generation time almost halved with a significant decrease of parser size
```bash
❯ time ts gen && du -h src/parser.c
tree-sitter gen  5,26s user 1,29s system 99% cpu 6,577 total
6,5M    src/parser.c
```
The same goes for other similar cases (e.g. catch/finally blocks in try_expression).
What might cause this degradation both in parser size and generation times? 
Are there any workarounds?

**Comment by eed3si9n** [comment, marked as answer]
I wrote about generation performance optimization a bit in https://eed3si9n.com/tree-sitter-scala-0.20.0, but a good entry point to figure out what could be happening is:

```bash
node_modules/.bin/tree-sitter generate --report-states-for-rule compilation_unit
```

As I understand it, as tree-sitter is expanding possible character scanning into a big state machine, you could get into a state explosion when you have lots of potential states that can funnel into your target state. For example, before sending https://github.com/tree-sitter/tree-sitter-scala/pull/102, I saw:

```bash
$  node_modules/.bin/tree-sitter generate --report-states-for-rule compilation_unit
class_definition                3728
function_definition             2214
ascription_expression           1442
infix_expression                1412
assignment_expression           1412
....
```

in other words, there were 3728 possible way a `class_definition` can be created, which is too many. Mainly I think this was caused by `prec.right(...)`. In other words, for almost any Scala code, you could add `class` in front of it and turn it into a class definition, but because it's right-associated, the parser constantly has to check if some random combination of token could be a legal class if you added `class` in front. The fix was to turn `class_definition` `proc.left(...)` and split up the right-associated part into `_class_constructor(...)`. So we still have right-association, but it would conditioned under first hitting `class` token.

I don't know if this intuition is correct, but overall applying the technique seems to have reduced the memory usage and time.

**Reply by susliko** [reply]
Hmm
I've done two tests: with left-associative and right-associative `$.class_definition` ([gist with results](https://gist.github.com/susliko/0c5086c42b4a7ef2a45033ae80bbf7f7))
They showed no difference in the number of states and used memory
Can it be that introduction of `$._class_constructor` gave the main contribution to the mentioned optimization?

**Reply by eed3si9n** [reply]
Yea. It's entirely possible that long chain of token is what's creating explosion in states, not left or right.
